sdk/core/Azure.Core/src/FailedResponseException.cs

```diff
@@ -2,33 +2,86 @@
 // Licensed under the MIT License.
 
 using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline.Policies;
 
 namespace Azure
 {
     public class RequestFailedException : Exception
     {
-        int _status;
+        private static string _defaultMessage = "Request failed with status code {0}";
 
-        public RequestFailedException(Response response, string summary = null)
-            : this(response, summary, null)
+        public int Status { get; }
+
+        public string Response { get; }
+
+        public RequestFailedException(int status, string message, string response)
+            : this(status, message, response, null)
         {}
 
-        public RequestFailedException(Response response, string summary, Exception innerException)
-            : base(FormatMessage(response, summary), innerException)
+        public RequestFailedException(int status, string message, string response, Exception innerException)
+            : base(message, innerException)
+        {
+            Status = status;
+            Response = response;
+        }
+
+        public override string ToString()
+        {
+            return base.ToString() + Environment.NewLine + Response;
+        }
+
+        public static Task<RequestFailedException> CreateAsync(Response response)
+        {
+            return CreateAsync(response, string.Format(_defaultMessage, response.Status));
+        }
+
+        public static Task<RequestFailedException> CreateAsync(Response response, string message)
         {
-            _status = response.Status;
+            return CreateAsync(message, response, true);
         }
 
-        private static string FormatMessage(Response response, string summary)
+        public static RequestFailedException Create(Response response)
         {
-            if (summary != null) {
-                return $"{summary}{Environment.NewLine}{response}";
+            return Create(response, string.Format(_defaultMessage, response.Status));
+        }
+
+        public static RequestFailedException Create(Response response, string message)
+        {
+            return CreateAsync(message, response, false).EnsureCompleted();
+        }
+
+        public static async Task<RequestFailedException> CreateAsync(string message, Response response, bool async)
```

Comment:
I think this should be on Response and then exception APIs should be all synchronous, just like any other exception.

Reply:
DO you want them as instance members or extensions?

Reply:
Feels strange to make them instance methods.